### PR TITLE
xcat-inventory support multiple object passwd enhancement

### PR DESCRIPTION
**Pull request by bybai**
xcat-inventory import passwd object task: https://github.com/xcat2/xcat-inventory/issues/138
xcat-inventory export passwd object issue: https://github.com/xcat2/xcat-inventory/issues/86

This PR realized : xcat-inventory import and export passwd object with 2 keys

UT:
1. export all passwd objects:
```
# xcat-inventory export -t passwd
passwd:
  omapi.xcat_key:
    password: xxxxxx
  xcat:
    cryptmethod: md5
    password: '12345'
  xcat.root:
    cryptmethod: md5
    password: '9999'
  xcat.wsuser:
    cryptmethod: md5
    password: '8888'
schema_version: '1.0'

#Version 2.14.4 (git commit 819b20b6602035d530c5d38afe6c79697151de20, built Fri Sep 21 06:17:52 EDT 2018)
```

2. export 'xcat.root' and 'xcat' objects
```
]# xcat-inventory export -t passwd -o xcat,xcat.root
passwd:
  xcat:
    cryptmethod: md5
    password: '12345'
  xcat.root:
    cryptmethod: md5
    password: '9999'
schema_version: '1.0'

#Version 2.14.4 (git commit 819b20b6602035d530c5d38afe6c79697151de20, built Fri Sep 21 06:17:52 EDT 2018)
```

3. export invalid object:
```
]# xcat-inventory export -t passwd -o xxxx
Error: cannot find passwd objects: xxxx!
```

4. export other object:
```
]# xcat-inventory export -t node -o test1
node:
  test1:
    device_type: switch
    obj_info:
      groups: all,test
    obj_type: node
    role: service
schema_version: '1.0'

#Version 2.14.4 (git commit 819b20b6602035d530c5d38afe6c79697151de20, built Fri Sep 21 06:17:52 EDT 2018)

]# xcat-inventory export -t site
schema_version: '1.0'
site:
  clustersite:
    SNsyncfiledir: /var/xcat/syncfiles
    auditnosyslog: '0'
    auditskipcmds: ALL
    blademaxp: '64'
    cleanupxcatpost: 'no'
    consoleondemand: 'no'
    databaseloc: /var/lib
    db2installloc: /mntdb2
    dhcplease: '43200'
    dnshandler: ddns
    domain: cluster.com
    enableASMI: 'no'
    forwarders: 10.5.106.2
    fsptimeout: '0'
    installdir: /install
    ipmimaxp: '64'
    ipmiretries: '3'
    ipmitimeout: '2'
    master: 10.5.106.7
    maxssh: '8'
    nameservers: 10.5.106.7
    nodesyncfiledir: /var/xcat/node/syncfiles
    powerinterval: '0'
    ppcmaxp: '64'
    ppcretry: '3'
    ppctimeout: '0'
    sharedtftp: '1'
    sshbetweennodes: ALLGROUPS
    syspowerinterval: '0'
    tftpdir: /tftpboot
    timezone: America/New_York
    useNmapfromMN: 'no'
    vsftp: n
    xcatconfdir: /etc/xcat
    xcatdport: '3001'
    xcatiport: '3002'
    xcatsslversion: TLSv1

#Version 2.14.4 (git commit 819b20b6602035d530c5d38afe6c79697151de20, built Fri Sep 21 06:17:52 EDT 2018)
```

5. import passwd objects:
```
[root@bybc0607 inventory]# cat /root/test
passwd:
  bai:
    cryptmethod: md5
    password: '666'
  bai.root:
    cryptmethod: md5
    password: '777'
schema_version: '1.0'

#Version 2.14.4 (git commit 819b20b6602035d530c5d38afe6c79697151de20, built Fri Sep 21 06:17:52 EDT 2018)
[root@bybc0607 inventory]# xcat-inventory import -f /root/test
loading inventory date in "/root/test"
start to import "passwd" type objects
 preprocessing "passwd" type objects
 writting "passwd" type objects
Inventory import successfully!
[root@bybc0607 inventory]# tabdump passwd
#key,username,password,cryptmethod,authdomain,comments,disable
"xcat",,"12345","md5",,,
"xcat","wsuser","8888","md5",,,
"omapi","xcat_key","xxxxxx",,,,
"xcat","root","9999","md5",,,
"bai","root","777","md5",,,
"bai",,"666","md5",,,
```

6. after clean up passwd table, import `/root/test`
```
[root@bybc0607 inventory]# xcat-inventory import -f /root/test -c
loading inventory date in "/root/test"
start to import "passwd" type objects
 preprocessing "passwd" type objects
 writting "passwd" type objects
Inventory import successfully!
[root@bybc0607 inventory]# tabdump passwd
#key,username,password,cryptmethod,authdomain,comments,disable
"bai","root","777","md5",,,
"bai",,"666","md5",,,
```

7. import node:
```
[root@bybc0607 inventory]# cat /root/test1
node:
  test2:
    device_type: server
    obj_info:
      groups: all,test
    obj_type: node
    role: compute
schema_version: '1.0'

[root@bybc0607 inventory]# xcat-inventory import -f /root/test1 -c
loading inventory date in "/root/test1"
start to import "node" type objects
 preprocessing "node" type objects
 writting "node" type objects
Inventory import successfully!
[root@bybc0607 inventory]# nodels
test2
[root@bybc0607 inventory]# lsdef test2
Object name: test2
    groups=all,test
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles
```

### CODE REVIEW

File: xcat-inventory/xcclient/inventory/manager.py [hunk 1]
```diff
@@ -301,9 +301,8 @@ def export_by_type(objtype, names, destfile=None, destdir=None, fmt='yaml',versi
     else:
         objtypelist.extend(InventoryFactory.getvalidobjtypes())
         exportall=1
-
     if names:
-        objlist.extend([n.strip() for n in names.split(',')])
+        objlist.extend([n.strip() and n.rstrip('.') for n in names.split(',')])
```
Comment (immarvin) [inline xcat-inventory/xcclient/inventory/manager.py]: `n.strip() and n.rstrip('.')` should better be `n.strip().rstrip('.')`

File: xcat-inventory/xcclient/inventory/dbobject.py [hunk 2]
```diff
@@ -47,6 +64,14 @@ class passwd(Base,mixin):
     __tablename__ = 'passwd'
     __table_args__ = {'autoload':True}
 
+    @classmethod
+    def getdictkey(cls):
+        return 'key.username'
+
+    @classmethod
+    def getkeys(cls):
```
Comment (immarvin) [inline xcat-inventory/xcclient/inventory/dbobject.py]: seems `mixin::getkeys` works for `passwd` class, no need to reload `getkeys ` here

File: xcat-inventory/xcclient/inventory/dbobject.py [hunk 3]
```diff
@@ -47,6 +64,14 @@ class passwd(Base,mixin):
     __tablename__ = 'passwd'
     __table_args__ = {'autoload':True}
 
+    @classmethod
+    def getdictkey(cls):
```
Comment (immarvin) [inline xcat-inventory/xcclient/inventory/dbobject.py]: should better add some comments for this function

File: xcat-inventory/xcclient/inventory/dbobject.py (hunk 2 above)
Comment (bybai) [reply xcat-inventory/xcclient/inventory/dbobject.py]: `getkeys()` in `passwd` object can keep keys order ['key','username']. The base `getkeys()` is general method.

**Comment by bybai** [comment]
`networks` and `switch` tables have multiple keys, I commit new enhancements for these 2 tables.
UT:
```
[root@bybc0607 xCAT_plugin]# xcat-inventory export -t node -o test2
node:
  test2:
    device_type: switch
    network_info:
      otherinterfaces: -bmc,10.5.5.5
      primarynic:
        ip: 10.4.4.4
        switch: switch
        switchinterface: switchinterface
        switchport: switchport
        switchvlan: switchvlan
    obj_info:
      groups: all,test
    obj_type: node
    role: service
schema_version: '1.0'

[root@bybc0607 xCAT_plugin]# xcat-inventory export -t network
network:
  10_0_0_0-255_0_0_0:
    basic_attr:
      gateway: 10.0.0.103
      mask: 255.0.0.0
      mgtifname: eth0
      mtu: '1500'
      net: 10.0.0.0
    service:
      tftpserver: <xcatmaster>
schema_version: '1.0'

#Version 2.14.4 (git commit 819b20b6602035d530c5d38afe6c79697151de20, built Fri Sep 21 06:17:52 EDT 2018)
```
```

**Comment by immarvin** [comment]
this fix in this PR will be covered by PR https://github.com/xcat2/xcat-inventory/pull/142, I would like to close  this PR after https://github.com/xcat2/xcat-inventory/pull/142 is merged

**Comment by bybai** [comment]
This fix is included in PR #142.  close this one.

**Pull request closed without merging**